[MRG+1] Affinity propagation edge cases (#9612)

jsamoocha opened the pull request:
#### Reference Issue
Fixes #9612


#### What does this implement/fix? Explain your changes.

1. `AffinityPropagation.predict(X)` would fail in case of non-convergence of the algorithm when fitting the model. It now returns the same label ('-1' for noise) for every sample in `X` in that case.
2. `AffinityPropagation.fit()` behavior was undefined and sometimes arbitrary (depending on preference and/or damping values) for cases when training samples had equal mutual similarities. It now behaves in a consistent way for these edge cases, i.e. returning one cluster when preference < mutual similarities and returning `n_samples` clusters when preference >= mutual similarities.

Commit:
Added test exposing non-convergence issues

As discussed in issue #9612, expecting cluster centers to be an empty array and labels to be
unique for every sample.

Commit:
Addresses non-convergence issues

Returns empty list as cluster center indices to prevent adding a dimension in fit() method,
returns unique labels for samples making this consistent with (TBD) predict() behavior for
non-convergence.

Commit:
Made predict() handle case of non-convergence while fitting

In this case, it will log a warning and return unique labels for every new sample.

Commit:
Added helper function for detecting mutually equal similarities and preferences

Commit:
Tidied imports

Commit:
Immediately returning trivial clusters and labels in case of equal similarities and preferences

Commit:
Simplified code for preference(s) equality test

Commit:
Corrected for failing unit tests covering case of n_samples=1

Commit:
Corrected for PEP8 line too long

Commit:
Rewriting imports to comply with max 80-column lines

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -327,4 +364,10 @@ def predict(self, X):
+        if self.cluster_centers_.size > 0:
+            return pairwise_distances_argmin(X, self.cluster_centers_)
+        else:
+            logger.warning("This model does not have any cluster centers "
```
For good or bad, we use warnings.warn, not logging...

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +109,23 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+        # cluster equal to the single sample
+        return (np.array([0]), np.array([0]), 0) if return_n_iter \
+            else (np.array([0], np.array([0])))
+    elif equal_similarities_and_preferences(S, preference):
```
Can't you just do if n_samples == 1 or equal... here?

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -177,9 +213,10 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
-        labels = np.empty((n_samples, 1))
-        cluster_centers_indices = None
-        labels.fill(np.nan)
+        logger.warning("Affinity propagation did not converge, this model "
```
Use a sklearn.exceptions.ConvergenceWarning

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +109,23 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+        # It makes no sense to run the algorithm in this case, so return 1 or
+        # n_samples clusters, depending on preferences
+        if np.array(preference).flat[0] >= S.flat[1]:
+            return (np.arange(n_samples), np.arange(n_samples), 0) \
```
Does this case deserve a warning?

jnothman reviewed (commented):
Great work, thanks!

Could you please test that/when a warning is raised with assert_warns? (In this PR or elsewhere we should also use a ConvergenceWarning in the existing non-convergence case...)

Commit:
Simplified code

n_samples == 1 case does not need a separate return statement.

Commit:
Replaced logging warnings by warnings.warn()

Added assertions for warnings in tests.

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +106,20 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+        warnings.warn("All samples have mutually equal similarities, "
+                      "returning arbitrary cluster center(s).")
+        if np.array(preference).flat[0] >= S.flat[n_samples - 1]:
+            return (np.arange(n_samples), np.arange(n_samples), 0) \
```
Aesthetics:
```
        if np.array(preference).flat[0] >= S.flat[n_samples - 1]:
            return (np.arange(n_samples), np.arange(n_samples), 0)
                    if return_n_iter
                    else (np.arange(n_samples), np.arange(n_samples))
```

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +106,20 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+        # n_samples clusters, depending on preferences
+        warnings.warn("All samples have mutually equal similarities, "
+                      "returning arbitrary cluster center(s).")
+        if np.array(preference).flat[0] >= S.flat[n_samples - 1]:
```
I think we can convert preference to an array outside of this condition and avoid repeatedly casting it...

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +106,20 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+    if n_samples == 1 or equal_similarities_and_preferences(S, preference):
+        # It makes no sense to run the algorithm in this case, so return 1 or
+        # n_samples clusters, depending on preferences
+        warnings.warn("All samples have mutually equal similarities, "
```
, -> .

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +106,20 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
+        # It makes no sense to run the algorithm in this case, so return 1 or
+        # n_samples clusters, depending on preferences
+        warnings.warn("All samples have mutually equal similarities, "
+                      "returning arbitrary cluster center(s).")
```
should this be a `ConvergenceWarning`? I suppose `UserWarning` is fine.

jsamoocha replied:
It is a `UserWarning` right now

jnothman commented on sklearn/cluster/tests/test_affinity_propagation.py:
```diff
@@ -78,3 +82,77 @@ def test_affinity_propagation_predict_error():
+    # Force non-convergence by allowing only a single iteration
+    af = AffinityPropagation(preference=-10, max_iter=1).fit(X)
+
+    assert_array_equal(np.array([0, 1, 2]), af.predict(X))
```
I'm afraid this doesn't make sense. We can't predict every sample as being in its own cluster in the inductive case, if this means that there are new clusters at predict time than at fit time. The implication here is also that some of the predicted items are clustered with the training points in the same position, which does not make sense.

Options:
* instead mark all points  as not clustered, i.e. label -1 as in dbscan
* raise an error in `predict` if there are no exemplars
* make each training point an exemplar

jsamoocha replied:
Good catch! That will prevent some confusion...

Commit:
Marking function as non-public

Commit:
Using mask instead of modifying S

Commit:
Improvement suggested by review comment

Commit:
Avoided casting preference to array twice

Commit:
Readability improvements

Commit:
Improved returned labels in case of no cluster centers

Returning a unique label for every sample in X suggests that these were based on actual clusters.
Since there are no clusters, it makes more sense to return a negative label for all samples,
indicating there were no clusters.

Commit:
PEP8 line too long

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -90,6 +106,23 @@ def affinity_propagation(S, preference=None, convergence_iter=15, max_iter=200,
     if damping < 0.5 or damping >= 1:
         raise ValueError('damping must be >= 0.5 and < 1')
 
+    preference_array = np.array(preference)
```
any reason not to reuse the name `preference`? It's quite clear from how it's used (e.g. `.flat`) that it must be an array.

jnothman commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -327,4 +361,10 @@ def predict(self, X):
+        else:
+            warnings.warn("This model does not have any cluster centers "
+                          "because affinity propagation did not converge. "
+                          "Returning unique labels for the provided samples.")
```
No longer the case.

Are you sure we don't just want to raise an error when a user tries to predict with such a model? Perhaps an array of -1 makes sense in fit_predict if we're going to do it here..?

jsamoocha replied:
`predict()` unexpectedly raising an error was the initial reason this PR got started. An error during `predict()` would be appropriate if the caller would provide incorrect data IMHO. Since these models can potentially live for a long time (in my use case, they're trained infrequently and are deserialized later with `joblib.load()`), I wouldn't want to make the caller of `predict()` responsible for dealing with potential crashes because of issues during training of the models. Returning -1 as labels (and logging the warning) seems to be a bit more friendly to the caller.

I'll address the incorrect warning message.

Commit:
Avoided creating separate variable for preference as array

Commit:
Corrected warning message

jnothman commented:
But then don't you think that labels_ should also be -1 then?

Commit:
Making labels consistent with predict() behavior in case of non-convergence

Commit:
Minor readability improvement

Commit:
Added detail to test comment about expected result

jsamoocha commented:
Definitely. I overlooked that part - busy with too many things at the same time :-/

jnothman reviewed (commented):
This is good, except that it deserves some documentation. Perhaps a Notes section in both class and function docstrings?

Commit:
Added documentation about edge cases

jnothman commented:
LGTM, thanks! Another review?

agramfort commented:
+1 for MRG

please just update what's new bug fix section and let's merge

jsamoocha commented:
Not too sure what it means to "update ... bug fix section", was any action expected from my side?

jnothman commented:
see doc/whats_new.rst

On 5 September 2017 at 17:08, Jonatan Samoocha <notifications@github.com>
wrote:

> Not too sure what it means to "update ... bug fix section", was any action
> expected from my side?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9635#issuecomment-327090279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_ky-O-0Siv3r4l7d7PCBsHKP_-Cks5sfPNzgaJpZM4PDtqs>
> .
>

Commit:
Added documentation to 'what's new'

jnothman commented:
Thanks!
